From f03716638c15a37d616075b190f8ee4c8e2d46fb Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Alex=20R=C3=B8nne=20Petersen?= <alex@alexrp.com>
Date: Wed, 29 Oct 2025 20:44:03 +0100
Subject: [PATCH 023/133] std.os.linux: fix restore for powerpc/powerpc64
X-Developer-Signature: v=1; a=openpgp-sha256; l=1966; i=xtex@astrafall.org;
 h=from:subject; bh=YV9C1OM172CgAF8zSF75RCvGkVdrG8b4g/xCDMEfNC0=;
 b=owGbwMvMwCW2U4Ij7wZL9ETG02pJDJn7cn8XZ5ct3Rln0vVS8PDmdb0bgrbs+3ecb2vl/Iw9X
 K9Xtnx/0DGRhUGMi8FSTJGlyLDBm1UnnV90WbkszBxWJpAh0iINDAwMDCwMfLmJeaVGOkZ6ptqG
 eoaGOkY6RgxcnAIw1dlVjAzz1fzfOcZpv/x8Ks5hm4D+h1kJyienGf53Tt4Yv+FjhfYURoaVpRb
 WVgu0QsNOWOS63156boq317IusUk2RwWTm+sn87ACAA==
X-Developer-Key: i=xtex@astrafall.org; a=openpgp;
 fpr=7231804B052C670F15A6771DB918086ED8045B91

sigreturn and rt_sigreturn are distinct syscalls on PowerPC.

(cherry picked from commit b2bc44e0d5e5edde083ec281aa0575b16478d881)
---
 lib/std/os/linux/powerpc.zig   | 16 +++++++++++++++-
 lib/std/os/linux/powerpc64.zig | 16 +++++++++++++++-
 2 files changed, 30 insertions(+), 2 deletions(-)

diff --git a/lib/std/os/linux/powerpc.zig b/lib/std/os/linux/powerpc.zig
index 94d7a7b8347c..87763b6515e1 100644
--- a/lib/std/os/linux/powerpc.zig
+++ b/lib/std/os/linux/powerpc.zig
@@ -244,7 +244,21 @@ pub fn clone() callconv(.naked) usize {
     );
 }
 
-pub const restore = restore_rt;
+pub fn restore() callconv(.naked) noreturn {
+    switch (builtin.zig_backend) {
+        .stage2_c => asm volatile (
+            \\ li 0, %[number]
+            \\ sc
+            :
+            : [number] "i" (@intFromEnum(SYS.sigreturn)),
+        ),
+        else => asm volatile (
+            \\ sc
+            :
+            : [number] "{r0}" (@intFromEnum(SYS.sigreturn)),
+        ),
+    }
+}
 
 pub fn restore_rt() callconv(.naked) noreturn {
     switch (@import("builtin").zig_backend) {
diff --git a/lib/std/os/linux/powerpc64.zig b/lib/std/os/linux/powerpc64.zig
index ed648a3a0aa7..cd2d927de1a5 100644
--- a/lib/std/os/linux/powerpc64.zig
+++ b/lib/std/os/linux/powerpc64.zig
@@ -229,7 +229,21 @@ pub fn clone() callconv(.naked) usize {
     );
 }
 
-pub const restore = restore_rt;
+pub fn restore() callconv(.naked) noreturn {
+    switch (builtin.zig_backend) {
+        .stage2_c => asm volatile (
+            \\ li 0, %[number]
+            \\ sc
+            :
+            : [number] "i" (@intFromEnum(SYS.sigreturn)),
+        ),
+        else => asm volatile (
+            \\ sc
+            :
+            : [number] "{r0}" (@intFromEnum(SYS.sigreturn)),
+        ),
+    }
+}
 
 pub fn restore_rt() callconv(.naked) noreturn {
     switch (@import("builtin").zig_backend) {
-- 
2.52.0

